internal/util/search/search.go

```diff
@@ -0,0 +1,178 @@
+package search
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/GoogleContainerTools/kpt/pkg/kptfile"
+	"sigs.k8s.io/kustomize/kyaml/errors"
+	"sigs.k8s.io/kustomize/kyaml/kio"
+	"sigs.k8s.io/kustomize/kyaml/kio/kioutil"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+const PathDelimiter = "."
+
+// SearchReplace struct holds the input parameters and results for
+// Search and Replace operations on resource configs
+type SearchReplace struct {
+	// Value is the value of the field to be matched
+	Value string
+
+	// ValueRegex is the value regex of the field to be matched
+	ValueRegex string
+
+	// Path is the path of the field to be matched
+	Path string
+
+	// Count is the number of matches
+	Count int
+
+	// Match is the map of file path to list of matched fields
+	Match map[string][]string
+
+	// PutLiteral is the literal to be put at to field
+	// filtered by path and/or value
+	PutLiteral string
+
+	filePath string
+}
+
+// Perform performs the search and replace operation on each node in the package path
+func (sr *SearchReplace) Perform(resourcesPath string) error {
+	inout := &kio.LocalPackageReadWriter{
+		PackagePath:     resourcesPath,
+		NoDeleteFiles:   true,
+		PackageFileName: kptfile.KptFileName,
+	}
+	sr.Match = make(map[string][]string)
+	return kio.Pipeline{
+		Inputs:  []kio.Reader{inout},
+		Filters: []kio.Filter{kio.FilterAll(sr)},
+		Outputs: []kio.Writer{inout},
+	}.Execute()
+}
+
+// Filter parses input node and performs search and replace operation on the node
+func (sr *SearchReplace) Filter(object *yaml.RNode) (*yaml.RNode, error) {
+	filePath, _, err := kioutil.GetFileAnnotations(object)
+	if err != nil {
+		return object, err
+	}
+	sr.filePath = filePath
+
+	if sr.shouldPutLiteralByPath() {
+		return object, sr.putLiteral(object)
+	}
+
+	// traverse the node to perform search/put operation
+	err = accept(sr, object)
+	return object, err
+}
+
+// visitMapping parses mapping node
+func (sr *SearchReplace) visitMapping(object *yaml.RNode, path string) error {
+	return nil
+}
+
+// visitSequence parses sequence node
+func (sr *SearchReplace) visitSequence(object *yaml.RNode, path string) error {
+	// TODO: pmarupaka support sequence nodes
+	return nil
+}
+
+// visitScalar parses scalar node
+func (sr *SearchReplace) visitScalar(object *yaml.RNode, path string) error {
+	regexMatch, err := sr.regexMatch(object.Document().Value)
+	if err != nil {
+		return err
+	}
+
+	pathMatch, err := sr.pathMatch(path)
+	if err != nil {
+		return err
+	}
+
+	valueMatch := object.Document().Value == sr.Value || regexMatch
+
+	// at least one of path or value must be matched
+	if (valueMatch && pathMatch) || (valueMatch && sr.Path == "") ||
+		(sr.Value == "" && sr.ValueRegex == "" && pathMatch) {
```

Comment:
Consistency, match test should come first:

(valueMatch && ...) ||
(pathMatch &&)

Also think about simplifying this conditional, can assert the negative instead (not matched)?

Reply:
Actually, a more readable way to structure this method:

```
if need_to_match_by_x then
  matched, err = match_by_x()
    if not(matched)
      return
if need_to_match_by_y
```

i.e. you want to return immediately if you know one of the predicates in the AND logic has already failed.